federatedrec/param/collaborative_memory_param.py

```diff
@@ -0,0 +1,316 @@
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+import copy
+import json
+import typing
+from types import SimpleNamespace
+
+from federatedml.param.base_param import BaseParam
+from federatedml.param.cross_validation_param import CrossValidationParam
+from federatedml.param.init_model_param import InitParam
+from federatedml.param.predict_param import PredictParam
+from federatedml.util import consts
+from federatedml.protobuf.generated import cmn_model_meta_pb2
+
+
+class CMNInitParam(InitParam):
+    def __init__(self, embed_dim=10, init_method='random_normal'):
+        super(CMNInitParam, self).__init__()
+        self.embed_dim = embed_dim
+        self.init_method = init_method
+
+    def check(self):
+        if type(self.embed_dim).__name__ not in ["int"] or self.embed_dim < 0:
+            raise ValueError(
+                "CMNInitParam's embed_dim {} not supported, should be 'int'"
+                "and greater than 0".format(
+                    self.embed_dim))
+        return True
+
+
+class CMNParam(BaseParam):
+    """
+    Parameters
+    ----------
+    optimizer : dict, support optimizers in Keras such as  'SGD', 'RMSprop', 'Adam',  or 'Adagrad',
+        default: 'SGD' with learning rate 0.01
+        Optimize method
+
+    batch_size : int, default: -1
+        Batch size when updating model. -1 means use all data in a batch. i.e. Not to use mini-batch strategy.
+
+    init_param: InitParam object, default: default InitParam object
+        Init param method object.
+
+    max_iter : int, default: 100
+        The maximum iteration for training.
+
+    early_stop : dict. early_stop includes 'diff', 'weight_diff' and 'abs',
+        default: {'early_stop':'diff', 'eps': 1e-5}
+        Method used to judge converge or not.
+            a)	diff： Use difference of loss between two iterations to judge whether converge.
+            b)  weight_diff: Use difference between weights of two consecutive iterations
+            c)	abs: Use the absolute value of loss to judge whether converge. i.e. if loss < eps, it is converged.
+
+
+    predict_param: PredictParam object, default: default PredictParam object
+
+    cv_param: CrossValidationParam object, default: default CrossValidationParam object
+
+    hops: hops count of CMN model
+
+    max_len: max length of neighbors
+
+    l2_coef: l2 penalty coefficient
+    """
+
+    def __init__(self,
+                 secure_aggregate: bool = True,
+                 aggregate_every_n_epoch: int = 1,
+                 early_stop: typing.Union[str, dict, SimpleNamespace] = {"early_stop": "diff"},
+                 optimizer: typing.Union[str, dict, SimpleNamespace] = {"optimizer": "SGD", "learning_rate": 0.01},
+                 batch_size=-1,
+                 init_param=CMNInitParam(),
+                 max_iter=100,
+                 predict_param=PredictParam(),
+                 cv_param=CrossValidationParam(),
```

Comment:
Cross validation param not used in CMN, consider remove extra params & param checker.